### DISCUSSION TITLE
Default UDP socket for user-traffic of a Subscriber [14236]

### DISCUSSION THREAD

**Discussion by JerrySciangula** [eProsima/Fast-DDS#2599] [Q&A]
Hi all,

imagine a scenario with a 2 DomainParticipants: 1 Publisher with multiple DataWriters publishing data over different Topics and 1 Subscriber, including multiple DataReaders, each subscribed to the DataWriters in a 1-to-1 fashion. 

The best-effort sending of DataWriters data is managed by the Asynchronous thread using the FlowController policy HIGH_PRIORITY. So, each DataWriter is associated to a certain priority and the output flow is ordered according to this priority.

Now, for the Subscriber, if you disable the SHM transport, FastDDS by default creates 3 threads listening on UDP sockets (2 for meta-traffic data and 1 for user-traffic).

Regarding data user-traffic: 

1) Is all the traffic (DataWriters' data) handled in FIFO order by the single UDP socket?
2) Is there a way to guarantee that DataWriters' data is received in the same order it has been sent?
3) Can I create multiple user-traffic UDP threads to differentiate and manage the prioritized user-traffic?

Thank you in advance

**Comment by JLBuenoLopez** [comment]
Hi @JerrySciangula,

> 1. Is all the traffic (DataWriters' data) handled in FIFO order by the single UDP socket?

Yes, the reception UDP socket is handled in FIFO order.

> 2. Is there a way to guarantee that DataWriters' data is received in the same order it has been sent?

This is only guaranteed in a per Topic basis. Fast DDS ensures that the samples received in a specific Topic are received in the same order as they have been sent. However, in a multi-topic scenario, this is not guaranteed. 

DDS specification includes within the [PresentationPolicyQos](https://fast-dds.docs.eprosima.com/en/latest/fastdds/dds_layer/core/policy/standardQosPolicies.html#presentationqospolicy) the `ordered_access` option which is not yet implemented in Fast DDS.

> 3. Can I create multiple user-traffic UDP threads to differentiate and manage the prioritized user-traffic?

If you set a custom listening locator to each DataReader, Fast DDS internally will create a single reception thread for each of them. This can be done using the [RTPSEndpointQos](https://fast-dds.docs.eprosima.com/en/latest/fastdds/dds_layer/core/policy/eprosimaExtensions.html#rtpsendpointqos) defining the locator within the `unicast_locator_list`.

**Reply by JerrySciangula** [reply]
Thank you very much for the clear and complete reply!
